### PR TITLE
Allow use of ServiceProvider func to create BlobClient for PersistKeysToAzureBlobStorage

### PR DESCRIPTION
**Pull request by thomhurst**
Allow use of ServiceProvider func to create BlobClient for PersistKeysToAzureBlobStorage call (for resolving nested dependencies)

I've come across code where we want to generate our blob client using other services (CertificateTokenCredential which then relies on a KeyVaultClient, etc.) and without access to the ServiceProvider, this is dependency hell.

Being able to use the ServiceProvider to resolve your blob client allows us to retrieve registered dependencies and write cleaner setup code where needed.

### CODE REVIEW

**Comment by jsquire** [comment]
Hi @thomhurst.   Thank you for your contributions.   Because this is a change to the public API, this is something that we'll need to carefully consider.  Would you please help me understand the scenario where the proposed API would be called.   Specifically, I'm interested in what problems it solves and what workarounds without the changes would entail.   Thanks.

**Comment by jsquire** [comment]
//fyi:  @JoshLove-msft

**Comment by asos-tomlonghurst** [comment]
@jsquire  and @JoshLove-msft - My TokenCredential I can't simply just 'new up' because it depends on other services to resolve it.
So I want to do:
```
services.AddDataProtection()
             .ProtectKeysWithAzureKeyVault(dataProtectionSettings.EncryptionKeyIdentifier, serviceProvider => 
                     serviceProvider.GetRequiredService<TokenCredential>())
```

vs what I'd currenty have to do:

```
var azureActiveDirectorySettings = new AzureActiveDirectorySettings();
var loggerSettings = new LoggerSettings();
var logger = new Logger(loggerSettings);
var certificateService = new CertificateService(azureActiveDirectorySettings, logger);
var azureActiveDirectoryCertificate = certificateService.FindByThumbprint(azureActiveDirectorySettings.AADCertificateThumbprint);

var tokenCredential = new ClientCertificateCredential(azureActiveDirectorySettings.AADTenantId, azureActiveDirectorySettings.AADApplicationId, azureActiveDirectoryCertificate);
```

Because there's no access to the ServiceProvider, I'd have to new up my services, creating ugly code, and I'm losing the benefits of dependency injection. I could also be creating new instances of a class that's meant to be a singleton.

**Comment by thomhurst** [comment]
@jsquire  and @JoshLove-msft - My TokenCredential I can't simply just 'new up' because it depends on other services to resolve it.
So I want to do:
```
services.AddDataProtection()
             .ProtectKeysWithAzureKeyVault(dataProtectionSettings.EncryptionKeyIdentifier, serviceProvider => 
                     serviceProvider.GetRequiredService<TokenCredential>())
```

vs what I'd currenty have to do:

```
var azureActiveDirectorySettings = new AzureActiveDirectorySettings();
var loggerSettings = new LoggerSettings();
var logger = new Logger(loggerSettings);
var certificateService = new CertificateService(azureActiveDirectorySettings, logger);
var azureActiveDirectoryCertificate = certificateService.FindByThumbprint(azureActiveDirectorySettings.AADCertificateThumbprint);

var tokenCredential = new ClientCertificateCredential(azureActiveDirectorySettings.AADTenantId, azureActiveDirectorySettings.AADApplicationId, azureActiveDirectoryCertificate);
```

Because there's no access to the ServiceProvider, I'd have to new up my services, creating ugly code, and I'm losing the benefits of dependency injection. I could also be creating new instances of a class that's meant to be a singleton.

**Comment by asos-tomlonghurst** [comment]
Sorry, realise the code example I supplied applies to this PR:
https://github.com/Azure/azure-sdk-for-net/pull/33454

But the same principal applies here for getting your blob client. Hope that makes sense!

**Comment by thomhurst** [comment]
Sorry, realise the code example I supplied applies to this PR:
https://github.com/Azure/azure-sdk-for-net/pull/33454

But the same principal applies here for getting your blob client. Hope that makes sense!

**Comment by jsquire** [comment]
Thanks, @thomhurst.   That visualization is helpful.   I'll start a conversation with our .NET architects internally to collect their feedback and provide an update here when available.  Please forgive us if that takes a couple of days to work through.

**Comment by thomhurst** [comment]
Thanks @jsquire !

**Comment by thomhurst** [comment]
@jsquire any updates internally?

**Comment by jsquire** [comment]
Not as of yet; to set expectations, it will likely take a few days to a week to get through the initial conversation due to the schedules of the folks needed.   As soon as we have something, I'll ping with an update.

**Comment by jsquire** [comment]
Hi @thomhurst.   Just checking-in to let you know that we've got the right stakeholders engaged and discussion is taking place internally.   I'm hopeful we'll have a final outcome this week and will update you as soon as we do.

**Comment by thomhurst** [comment]
Thanks for the update @jsquire  :)

**Review by JoshLove-msft — APPROVED** [review]

**Comment by thomhurst** [comment]
@JoshLove-msft @jsquire  How long would something like this take to be released in a new package version?

File: sdk/extensions/Azure.Extensions.AspNetCore.DataProtection.Blobs/src/ConfigureKeyManagementBlobClientOptions.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using Microsoft.AspNetCore.DataProtection.KeyManagement;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
+
+#pragma warning disable AZC0001 //
```
Comment (jsquire) [inline sdk/extensions/Azure.Extensions.AspNetCore.DataProtection.Blobs/src/ConfigureKeyManagementBlobClientOptions.cs:8]: nit:  Please leave a short description of the rule being suppressed.

**Review by jsquire — APPROVED** [review]

**Comment by jsquire** [comment]
Hi @thomhurst.   Though it's likely obvious from the review activity, after internal discussion we believe the proposed changes help to improve the usability of the library and we'd like to move forward with them.   Thank you, again, for your kind contribution.

File: sdk/extensions/Azure.Extensions.AspNetCore.DataProtection.Blobs/src/ConfigureKeyManagementBlobClientOptions.cs (hunk 1 above)
Comment (thomhurst) [reply sdk/extensions/Azure.Extensions.AspNetCore.DataProtection.Blobs/src/ConfigureKeyManagementBlobClientOptions.cs]: Wasn't necessary - Builds without any warnings even without. Have removed

**Comment by jsquire** [comment]
Hi @thomhurst.   Just looping back around to let you know that both packages were released today and the latest version on NuGet should contain these changes.   Thank you, again, for your contributions.

**Pull request merged**